### PR TITLE
Sdl client cpp

### PR DESCRIPTION
**Pull request by akallabeth**
Migrate `sdl-client` codebase, use `C++` for client implementation

### CODE REVIEW

File: client/SDL/CMakeLists.txt [hunk 1]
```diff
@@ -29,31 +55,49 @@ if (WITH_DEBUG_SDL_KBD_EVENTS)
...
+	include_directories(${FreeRDP_INCLUDE_DIR})
+
+	find_package(FreeRDP-Client REQUIRED)
+	include_directories(${FreeRDP-Client_INCLUDE_DIR})
+endif()
+
 set(SRCS
-    sdl_utils.c
+    sdl_utils.cpp
     sdl_utils.h
```
Comment (pnowack) [inline client/SDL/CMakeLists.txt:76]: I think the header files could be adjusted too (`.h` -> `.hpp`).

File: client/SDL/sdl_freerdp.cpp [hunk 2]
```diff
@@ -303,27 +305,43 @@ static BOOL sdl_redraw(sdlContext* sdl)
 	return gdi_send_suppress_output(gdi, FALSE);
 }
 
+class EventGuard
```
Comment (hardening) [inline client/SDL/sdl_freerdp.cpp:308]: The class name seem way too generic while it's doing something very specific

File: client/SDL/CMakeLists.txt (hunk 1 above)
Comment (akallabeth) [reply client/SDL/CMakeLists.txt]: done

File: client/SDL/sdl_freerdp.cpp (hunk 2 above)
Comment (akallabeth) [reply client/SDL/sdl_freerdp.cpp:308]: it is a private class, so no need to be universally unique.

**Comment by akallabeth** [comment]
@pnowack @hardening this pr is ready (e.g. now builds fine on all platforms), so please review :)

File: client/SDL/sdl_freerdp.cpp (hunk 2 above)
Comment (akallabeth) [reply client/SDL/sdl_freerdp.cpp]: ok, renamed.

File: client/SDL/sdl_disp.cpp [hunk 3]
```diff
@@ -248,14 +250,14 @@ sdlDispContext* sdl_disp_new(sdlContext* sdl)
 	rdpSettings* settings;
 
 	if (!sdl || !sdl->common.context.settings || !sdl->common.context.pubSub)
-		return NULL;
+		return nullptr;
 
 	settings = sdl->common.context.settings;
 	pubSub = sdl->common.context.pubSub;
-	ret = calloc(1, sizeof(sdlDispContext));
+	ret = new (sdlDispContext);
```
Comment (pnowack) [inline client/SDL/sdl_disp.cpp]: The braces (`()`) are unneeded.

File: client/SDL/sdl_disp.cpp [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+	if (_lastSentDesktopOrientation != settings->DesktopOrientation)
+		return TRUE;
+
+	if (_lastSentDesktopScaleFactor != settings->DesktopScaleFactor)
+		return TRUE;
+
+	if (_lastSentDeviceScaleFactor != settings->DeviceScaleFactor)
+		return TRUE;
+	/* TODFO
```
Comment (pnowack) [inline client/SDL/sdl_disp.cpp:57]: Typo: `TODFO` -> `TODO`

File: client/SDL/sdl_disp.cpp [hunk 5]
```diff
@@ -386,151 +297,132 @@ BOOL sdl_disp_handle_display_event(sdlDispContext* disp, const SDL_DisplayEvent*
...
+	_lastSentWidth = _targetWidth = settings->DesktopWidth;
+	_lastSentHeight = _targetHeight = settings->DesktopHeight;
+	PubSub_SubscribeActivated(pubSub, sdlDispContext::OnActivated);
+	PubSub_SubscribeGraphicsReset(pubSub, sdlDispContext::OnGraphicsReset);
+	PubSub_SubscribeTimer(pubSub, sdlDispContext::OnTimer);
+}
+
+sdlDispContext::~sdlDispContext()
+{
+	if (_sdl)
```
Comment (pnowack) [inline client/SDL/sdl_disp.cpp:424]: We already have an assertion for this in the body of the constructor, so this is unneeded (always true).

File: client/SDL/sdl_freerdp.cpp [hunk 6]
```diff
@@ -305,16 +305,16 @@ static BOOL sdl_redraw(sdlContext* sdl)
 	return gdi_send_suppress_output(gdi, FALSE);
 }
 
-class EventGuard
+class SdlEventUpdateTriggerGuard
```
Comment (pnowack) [inline client/SDL/sdl_freerdp.cpp:308]: Could this fixup commit be squashed into the first commit here?

**Review by pnowack — APPROVED** [review]
Just some nits, couldn't spot an issue here. Very nice RAII usage!

File: client/SDL/sdl_freerdp.cpp (hunk 6 above)
Comment (akallabeth) [reply client/SDL/sdl_freerdp.cpp:308]: done

File: client/SDL/sdl_disp.cpp (hunk 4 above)
Comment (akallabeth) [reply client/SDL/sdl_disp.cpp:57]: done

**Pull request merged**